Svelte 5: Type issues with derived store / page store

### Describe the bug

Just upgraded to Svelte 5 and I'm getting type errors for some stores-only code (`.ts` file).

### Reproduction

```ts
import { derived, readable, writable, type Updater, type Writable } from "svelte/store"
import { page } from "$app/stores"
import { browser } from "$app/environment"
import deepEquals from "fast-deep-equal"

let oldState: App.PageState = {}

export const pageState = derived<typeof page, App.PageState>(
    page,
    (_, set) => {
        if (!browser) return
        setTimeout(() => {
            const newState = history.state["sveltekit:states"] ?? {}
            if (!deepEquals(oldState, newState)) {
                oldState = newState
                set(newState)
            }
        })
    },
    {},
)
```

#### Workaround

Manually type the arrow function - replace `_, set` with:

```ts
_: typeof page extends Readable<infer U> ? U : never, set: (value: App.PageState) => void)
```

### Logs

```shell
$ tsc --noEmit true
src/lib/stores.ts:13:6 - error TS7006: Parameter '_' implicitly has an 'any' type.

13     (_, set) => {
        ~

src/lib/stores.ts:13:9 - error TS7006: Parameter 'set' implicitly has an 'any' type.

13     (_, set) => {
```


### System Info

```shell
System:
    OS: Windows 10 10.0.19045
    CPU: (20) x64 12th Gen Intel(R) Core(TM) i7-12700KF
    Memory: 13.10 GB / 31.86 GB
  Binaries:
    Node: 20.8.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 4.1.1 - C:\Program Files\nodejs\yarn.CMD
    npm: 10.2.1 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: Spartan (44.19041.1266.0)
    Internet Explorer: 11.0.19041.1566
```


### Severity

annoyance

--- a/.changeset/giant-bananas-turn.md
+++ b/.changeset/giant-bananas-turn.md
@@ -1,0 +1,5 @@
+---
+"svelte": patch
+---
+
+fix: adjust order of `derived` function definition overloads

--- a/packages/svelte/src/store/index.js
+++ b/packages/svelte/src/store/index.js
@@ -111,7 +111,7 @@ export function writable(value, start = noop) {
  * @template T
  * @overload
  * @param {S} stores
- * @param {(values: import('./private.js').StoresValues<S>) => T} fn
+ * @param {(values: import('./private.js').StoresValues<S>, set: (value: T) => void, update: (fn: import('./public.js').Updater<T>) => void) => import('./public.js').Unsubscriber | void} fn
  * @param {T} [initial_value]
  * @returns {import('./public.js').Readable<T>}
  */
@@ -124,7 +124,7 @@ export function writable(value, start = noop) {
  * @template T
  * @overload
  * @param {S} stores
- * @param {(values: import('./private.js').StoresValues<S>, set: (value: T) => void, update: (fn: import('./public.js').Updater<T>) => void) => import('./public.js').Unsubscriber | void} fn
+ * @param {(values: import('./private.js').StoresValues<S>) => T} fn
  * @param {T} [initial_value]
  * @returns {import('./public.js').Readable<T>}
  */

--- a/packages/svelte/tests/store/test.ts
+++ b/packages/svelte/tests/store/test.ts
@@ -272,7 +272,6 @@ describe('derived', () => {
 		const number = writable(1);
 		const evens = derived(
 			number,
-			// @ts-expect-error TODO feels like inference should work here
 			(n, set) => {
 				if (n % 2 === 0) set(n);
 			},
@@ -303,10 +302,8 @@ describe('derived', () => {
 		const number = writable(1);
 		const evensAndSquaresOf4 = derived(
 			number,
-			// @ts-expect-error TODO feels like inference should work here
 			(n, set, update) => {
 				if (n % 2 === 0) set(n);
-				// @ts-expect-error TODO feels like inference should work here
 				if (n % 4 === 0) update((n) => n * n);
 			},
 			0
@@ -442,7 +439,6 @@ describe('derived', () => {
 		const values: number[] = [];
 		const cleaned_up: number[] = [];
 
-		// @ts-expect-error TODO feels like inference should work here
 		const d = derived(num, ($num, set) => {
 			set($num * 2);
 
@@ -516,7 +512,6 @@ describe('derived', () => {
 		const a = writable(true);
 		let b_started = false;
 
-		// @ts-expect-error TODO feels like inference should work here
 		const b = derived(a, (_, __) => {
 			b_started = true;
 			return () => {
@@ -525,7 +520,6 @@ describe('derived', () => {
 			};
 		});
 
-		// @ts-expect-error TODO feels like inference should work here
 		const c = derived(a, ($a, set) => {
 			if ($a) return b.subscribe(set);
 		});

--- a/packages/svelte/tests/types/store.ts
+++ b/packages/svelte/tests/types/store.ts
@@ -11,3 +11,17 @@ derived([a], ([aVal]) => {
 	aVal === '';
 	return aVal === true;
 });
+
+derived(
+	a,
+	(value, set) => {
+		set('works');
+		// @ts-expect-error
+		set(true);
+
+		value === true;
+		// @ts-expect-error
+		value === '';
+	},
+	''
+);

--- a/packages/svelte/types/index.d.ts
+++ b/packages/svelte/types/index.d.ts
@@ -2157,14 +2157,14 @@ declare module 'svelte/store' {
 	 *
 	 * https://svelte.dev/docs/svelte-store#derived
 	 * */
-	export function derived<S extends Stores, T>(stores: S, fn: (values: StoresValues<S>) => T, initial_value?: T | undefined): Readable<T>;
+	export function derived<S extends Stores, T>(stores: S, fn: (values: StoresValues<S>, set: (value: T) => void, update: (fn: Updater<T>) => void) => Unsubscriber | void, initial_value?: T | undefined): Readable<T>;
 	/**
 	 * Derived value store by synchronizing one or more readable stores and
 	 * applying an aggregation function over its input values.
 	 *
 	 * https://svelte.dev/docs/svelte-store#derived
 	 * */
-	export function derived<S extends Stores, T>(stores: S, fn: (values: StoresValues<S>, set: (value: T) => void, update: (fn: Updater<T>) => void) => Unsubscriber | void, initial_value?: T | undefined): Readable<T>;
+	export function derived<S extends Stores, T>(stores: S, fn: (values: StoresValues<S>) => T, initial_value?: T | undefined): Readable<T>;
 	/**
 	 * Takes a store and returns a new one derived from the old one that is readable.
 	 *

